File: python/automarketing/ai/openai_compat.py
"""
OpenAI-format valaa provider (Groq, OpenRouter, ane biju ghana).

Aa badha ek j `/chat/completions` format vaapre che, etle ek j class thi
kaam thai jaay che. Navo provider umervo hoy to fakt base URL, model ane
key aapo — biju kai lakhvanu nathi.
"""

from __future__ import annotations

from typing import Any, Optional

from ..errors import FatalError, NotConfigured, RetryableError
from ..pipeline.http import get_client, parse_json_loose
from .base import (
    CompletionRequest,
    PreparedImage,
    TextProvider,
    VisionRequest,
    schema_to_prompt,
)


class OpenAiCompatProvider(TextProvider):
    """OpenAI na format valo koi pan provider."""

    def __init__(
        self,
        *,
        key: str,
        label: str,
        base_url: str,
        api_key: callable,
        model: callable,
        vision_model: Optional[callable] = None,
        hint: str,
        extra_headers: Optional[callable] = None,
        free: bool = True,
    ) -> None:
        self.key = key
        self.label = label
        self.free = free
        self._base_url = base_url.rstrip("/")
        self._api_key = api_key
        self._model = model
        self._vision_model = vision_model
        self._hint = hint
        self._extra_headers = extra_headers
        self.supports_vision = vision_model is not None

    @property
    def model(self) -> str:
        return self._model()

    def configured(self) -> bool:
        return bool(self._api_key())

    def missing_key_hint(self) -> str:
        return self._hint

    # ---------------------------------------------------------------- #

    def _headers(self) -> dict[str, str]:
        headers = {
            "content-type": "application/json",
            "authorization": f"Bearer {self._api_key()}",
        }
        if self._extra_headers:
            headers.update(self._extra_headers())
        return headers

    async def _chat(self, payload: dict[str, Any], *, timeout: float) -> str:
        if not self._api_key():
            raise NotConfigured(self._hint)

        client = get_client()
        response = await client.post(
            f"{self._base_url}/chat/completions",
            headers=self._headers(),
            json=payload,
            timeout=timeout,
        )

        if response.status_code >= 400:
            body = response.text[:500]

            if response.status_code == 429:
                raise RetryableError(
                    f"{self.label} ni rate limit lagi. Thodi var pachi apoaap "
                    f"fari try thashe. ({body[:160]})"
                )
            if response.status_code in (401, 403):
                raise FatalError(f"{self.label}: key khoti che ke band thai gai. ({body[:160]})")
            if response.status_code >= 500:
                raise RetryableError(f"{self.label}: server ni bhool ({response.status_code})")
            raise FatalError(f"{self.label}: HTTP {response.status_code} {body[:200]}")

        data = response.json()

        # Ketlak provider 200 sathe pan error andar mokle che.
        if isinstance(data, dict) and data.get("error"):
            message = str(data["error"].get("message") or data["error"])
            raise RetryableError(f"{self.label}: {message[:200]}")

        choices = (data or {}).get("choices") or []
        if not choices:
            raise RetryableError(f"{self.label} e khali jawab aapyo")

        content = (choices[0].get("message") or {}).get("content")
        if not content:
            raise RetryableError(f"{self.label} e khali jawab aapyo")

        return content

    # ---------------------------------------------------------------- #

    async def complete(self, request: CompletionRequest) -> Any:
        # Free model badha `json_schema` nathi samajta, etle schema prompt
        # ma j aapiye chie ane `json_object` mode maagie chie. Aa badhe
        # chale che ane jawab pan barabar aave che.
        payload = {
            "model": self.model,
            "messages": [
                {"role": "system", "content": f"{request.system}\n\n{schema_to_prompt(request.schema)}"},
                {"role": "user", "content": request.prompt},
            ],
            "response_format": {"type": "json_object"},
            "max_tokens": request.max_tokens,
            "temperature": request.temperature,
        }
        text = await self._chat(payload, timeout=120.0)
        return parse_json_loose(text, label=self.label)

    async def see(self, images: list[PreparedImage], request: VisionRequest) -> Any:
        if not self._vision_model:
            raise NotImplementedError(f"{self.key} image nathi joi shakto")

        content: list[dict[str, Any]] = [
            {
                "type": "image_url",
                "image_url": {"url": f"data:{img.mime_type};base64,{img.base64}"},
            }
            for img in images
        ]
        content.append({"type": "text", "text": request.prompt})

        payload = {
            "model": self._vision_model(),
            "messages": [
                {"role": "system", "content": f"{request.system}\n\n{schema_to_prompt(request.schema)}"},
                {"role": "user", "content": content},
            ],
            "response_format": {"type": "json_object"},
            "max_tokens": request.max_tokens,
            "temperature": request.temperature,
        }
        text = await self._chat(payload, timeout=150.0)
        return parse_json_loose(text, label=self.label)

File: src/lib/ai/openai-compat.ts
import { AiError, type CompletionRequest, type ProviderKey } from "./types";

/**
 * Ghana providers (OpenRouter, Groq, Together, DeepSeek, Mistral, local
 * llama.cpp...) OpenAI nu j `/chat/completions` format vapre che. Etle ek j
 * factory thi badha bani jay che — navu provider umervu hoy to fakt base URL
 * ane model naam aapo.
 */

export type OpenAiCompatConfig = {
  key: ProviderKey;
  label: string;
  free: boolean;
  baseUrl: string;
  apiKey: () => string;
  model: () => string;
  /** OpenRouter ne aa headers game che. */
  extraHeaders?: () => Record<string, string>;
  /** Aa provider `json_schema` response_format samje che? */
  supportsJsonSchema?: boolean;
  /** Key nathi tyare su kehvu. */
  missingKeyHint: string;
};

export function createOpenAiCompatProvider(config: OpenAiCompatConfig) {
  async function request<T>(
    req: CompletionRequest,
    useSchema: boolean,
  ): Promise<T> {
    const key = config.apiKey();
    if (!key) throw new AiError(config.missingKeyHint, config.key);

    const responseFormat = useSchema
      ? {
          type: "json_schema" as const,
          json_schema: {
            name: "response",
            strict: true,
            schema: req.schema,
          },
        }
      : { type: "json_object" as const };

    // json_object mode ma model ne schema prompt ma j batavvo pade che.
    const systemText = useSchema
      ? req.system
      : `${req.system}\n\nReply with ONLY a JSON object matching this JSON Schema (no markdown fence, no commentary):\n${JSON.stringify(req.schema)}`;

    let response: Response;
    try {
      response = await fetch(`${config.baseUrl}/chat/completions`, {
        method: "POST",
        headers: {
          "content-type": "application/json",
          authorization: `Bearer ${key}`,
          ...(config.extraHeaders?.() ?? {}),
        },
        body: JSON.stringify({
          model: config.model(),
          messages: [
            { role: "system", content: systemText },
            { role: "user", content: req.prompt },
          ],
          response_format: responseFormat,
          max_tokens: req.maxTokens ?? 4000,
          temperature: 0.8,
        }),
        signal: AbortSignal.timeout(120_000),
      });
    } catch (error) {
      throw new AiError(
        `${config.label} sudhi pahonchi na shakaya: ${(error as Error).message}`,
        config.key,
        true,
      );
    }

    const json = (await response.json()) as {
      choices?: Array<{ message?: { content?: string } }>;
      error?: { message?: string; code?: string };
    };

    if (!response.ok || json.error) {
      const message = json.error?.message ?? `HTTP ${response.status}`;

      if (response.status === 429) {
        throw new AiError(
          `${config.label} ni rate limit lagi. Thodi var pachi apoaap fari try thashe.`,
          config.key,
          true,
        );
      }
      if (response.status === 401 || response.status === 403) {
        throw new AiError(`${config.label}: key khoti che`, config.key);
      }
      // json_schema support na hoy to json_object thi fari try karo.
      if (useSchema && /schema|response_format|not support/i.test(message)) {
        return request<T>(req, false);
      }
      throw new AiError(
        `${config.label}: ${message}`,
        config.key,
        response.status >= 500,
      );
    }

    const text = json.choices?.[0]?.message?.content;
    if (!text) {
      throw new AiError(`${config.label} e khali jawab aapyo`, config.key, true);
    }

    return parseJsonLoose<T>(text, config.key, config.label);
  }

  return {
    key: config.key,
    label: config.label,
    free: config.free,
    get model() {
      return config.model();
    },
    configured() {
      return Boolean(config.apiKey());
    },
    complete<T>(req: CompletionRequest): Promise<T> {
      return request<T>(req, config.supportsJsonSchema !== false);
    },
  };
}

/**
 * Nana models kyarek ```json fence ke aagal-pachhal lakhan umeri de che.
 * Sadho JSON.parse fail thay to pehla `{` thi chhella `}` sudhi kaadhi laiye.
 */
export function parseJsonLoose<T>(
  text: string,
  provider: ProviderKey,
  label: string,
): T {
  const cleaned = text
    .trim()
    .replace(/^```(?:json)?\s*/i, "")
    .replace(/\s*```$/, "")
    .trim();

  try {
    return JSON.parse(cleaned) as T;
  } catch {
    const start = cleaned.indexOf("{");
    const end = cleaned.lastIndexOf("}");
    if (start >= 0 && end > start) {
      try {
        return JSON.parse(cleaned.slice(start, end + 1)) as T;
      } catch {
        /* niche error aapiye */
      }
    }
    throw new AiError(`${label} no jawab JSON ma nathi`, provider, true);
  }
}
